fvm/src/kernel/mod.rs

```diff
@@ -209,6 +209,14 @@ pub trait ActorOps {
         delegated_address: Option<Address>,
     ) -> Result<()>;
 
+    fn upgrade_actor<K: Kernel>(
+        &mut self,
+        new_code_cid: Cid,
+        params_id: BlockId,
+    ) -> Result<SendResult>;
+
+    fn can_actor_upgrade(&self) -> bool;
```

Comment:
IMO, this should be an implementation detail of the specific kernel, not something exposed on the interface. I.e., don't call it in the syscall binding, unconditionally call the `upgrade_actor` method on the kernel and check this precondition internally.